Generate flag values on startup.

komish opened the pull request:
Fixes #4 
Depends on #19 (when merged, 4 of these commits should go away, if they don't, I'll rebase).

This PR reconfigures the flag package a bit to allow it to be integrated into the workflow. This means that flag _values_ are generated on startup based on an input seed and the value of the `Template` key.

To do this, I needed to remove the value of the Flag that was hardcoded in flags.json.

We still don't have a way to accept a seed value, but that's not too far off. Same with adding the validation endpoint (see #3).

To avoid naming collisions I used the term "challenges" to represent the task that returns a flag to the user. Happy to change that, just couldn't figure out how to avoid a conflict with the `flag` package.

komish committed:
enable golangci-lint with all linters except unused and bump version

Signed-off-by: Jose R. Gonzalez <jose@flutes.dev>

komish committed:
generate flag values at launch based on seed

Signed-off-by: Jose R. Gonzalez <jose@flutes.dev>

tonyskapunk committed:
Initial work on generating flags

tonyskapunk commented:
:wave: @komish thanks for helping here. I'll take a look at your changes.

I had some initial work on this, but I never committed :crying_cat_face:
I pushed it to the [yctf/flags](https://github.com/tonyskapunk/yctf/tree/flags) branch where I had to make some changes you mentioned regarding the hardcoded flags. I'm OK dropping my work, but maybe there's something that could be reused.

tonyskapunk commented on flags.json:
```diff
@@ -1,47 +1,38 @@
 [
     {
        "ID": 0,
-       "Flag": "yctf-5f39410913126cea50a3e6fcdb7312628a869da376985fdb7ef7d0a41254a0e3", 
```
In the flags branch, my idea is to completely getting rid of the flags.json, while the file is there it's not used any longer

komish replied:
One thing to think about here is that the flags.json binds an identifier with its template. I don't think I saw an order enforced in the flags branch, but I know it's a WIP. In my mind, I think having some degree of configuration whereby you bind a specific template with a specific ID is okay.

tonyskapunk commented on main.go:
```diff
@@ -24,6 +29,37 @@ func main() {
+	flag.SetFactory(flag.NewFactory(defaultFlagFactorySeed))
+}
+
+func generateFlags() {
```
In the flags branch I'm generating the flags based on the session seed, this way

https://github.com/tonyskapunk/yctf/blob/511be24f3ddb3fb2a352809638f70b793d6dc981/models.go#L51-L60

it's important for me that the flag is not always the same, thus having the session as a seed is perfect as it will be always different per-user. Additionaly it helps with #3 and also to what was discussed in #10

komish replied:
That makes sense. That said, I would say that it's worth making sure there's some degree of repeatability. The specific case I have in mind is a situation where a server crashes in a local CTF environment. It would probably be best if administrators could ensure that completed flags for a participant stays completed verifiable on the restart of the server.

tonyskapunk reviewed (commented):
Thanks for all this work. I think we can use some parts from here and from what I have in the flags branch. I would like to retain the removal of flags.json and the creation of flags based on a seed. I'll get that ready after the integration of the other PRs, #18  and #19 

From here, at the very least we could retain the changes to the flag factory.

komish commented:
@tonyskapunk thanks for the review. Responses submitted. It sounds like you'll pull in the flags branch and then cherry pick whatever bits you need from here, so I'll assume this PR will close without merge. Let me know if you need anything! 👍

tonyskapunk commented:
> It sounds like you'll pull in the flags branch and then cherry pick whatever bits you need from here, so I'll assume this PR will close without merge.

Not sure yet; I may reuse this PR and bisect from mine and yours; I'll work on some of it this weekend. Thanks!

tonyskapunk committed:
Remove challenges

For now removing them, the idea is good and likely will get it back, one
of the main targets is to to use the session as the seed.

tonyskapunk reviewed (approved):
Thanks again for the help; I took out the "challenges" part for now, but most likely, I'll add something similar back. I want to make the session the seed for the flags and add the login + the score in place.
